man/cssplit.1
.TH CSSPLIT 1 "16 November 2025"
.\"
.\" (C) Copyright 2025 Diomidis Spinellis
.\"
.\" This file is part of CScout.
.\"
.\" CScout is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" CScout is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with CScout.  If not, see <http://www.gnu.org/licenses/>.
.\"
.SH NAME
cssplit \- Split a CScout processing script into smaller scripts
.SH SYNOPSIS
.B cssplit
.RB [ \-h ]
.B \-s
.I N
.I file
.RI [ file\ ... ]
.SH DESCRIPTION
The \fIcssplit\fP(1) command-line program splits a \fICScout\fP
processing script into \fIN\fP smaller scripts for parallel
or distributed processing.
This is useful for distributing the workload of a large \fICScout\fP
analysis across multiple scripts for efficiency or manageability.
Split scripts can be used to speed-up an initial check
of the processing with \fCcscout -c\fP
or to create a set of \fISQLite\fP database files that
are then merged with \fIcsmerge\fP(1).
.PP
The input is read from the specified file(s),
and the output files are written into the current working directory.
The input file(s) must have been created with \fIcsmake\fP(1).
Each output script is named using a zero-padded sequence,
starting with \fIfile-0000.cs\fP.
.SH OPTIONS
.TP
.BR \-s ", " \-\-shards " " N
Specifies the number of output scripts to generate.
The input processing script is divided into \fIN\fP approximately equal parts.
Each output script contains the appropriate subset of the input commands to
ensure no loss of information during the split.
Compilation units from multiple inter-dependent projects
(e.g. different configurations of the same executable)
are distributed to the scripts so as to maximize locality of reference.

.TP
.BR \-h ", " \-\-help
Display a brief usage message and exit.
.SH EXAMPLE
To split an input processing script into 4 smaller scripts:
.PP
.DS
.ft C
.nf
cssplit -s 4  make.cs
.ft P
.fi
.DE
.PP
This will create the following files in the current directory:
.PP
.DS
.ft C
.nf
file-0000.cs
file-0001.cs
file-0002.cs
file-0003.cs
.ft P
.fi
.DE
.PP
Each file contains a portion of the original processing script.
.SH "SEE ALSO"
\fIcscout\fP(1),
\fIcscut\fP(1),
\fIcsmake\fP(1),
\fIcsmerge\fP(1),
\fIpython\fP(1)
.SH AUTHOR
(C) Copyright 2024-2025 Diomidis Spinellis.


src/cssplit.py
#!/usr/bin/env python3
"""
Split CScout processing file into multiple processing shards.
"""

import sys
from pathlib import Path
from dataclasses import dataclass, field
from typing import List, TextIO
import argparse
import re
import signal

@dataclass
class CUProps:
    path: str
    offsets: List[int] = field(default_factory=list)
    allocated: bool = False

    def number_of_projects(self):
        """Return number of projects in which this CU belongs."""
        return sum(1 for off in self.offsets if off > 0)

    def add_project(self, project_id: int, offset):
        """Add the CU to the specified offset in project."""
        if project_id >= len(self.offsets):
            self.offsets.extend([0] * (project_id + 1 - len(self.offsets)))
        self.offsets[project_id] = offset

    def in_project(self, project_id: int):
        """Return true if the CU belongs to the specified project."""
        return project_id < len(self.offsets) and self.offsets[project_id] > 0

    def write(self, output_file: TextIO, project_id: int):
        """Write out the specified compilation unit for the given
        project."""
        input_file = input_files[project_id]
        input_file.seek(self.offsets[project_id], 0)
        for line in input_file:
            output_file.write(line)
            if line.startswith('#pragma echo "Done processing /'):
                return

project_name_to_id: dict[str, int] = {}
project_id_to_name: dict[int, str] = {}

project_id = -1

# Total number of compilation units
ncus = 0

cu_to_props: dict[str, CUProps] = {}

cu_list: list[CUProps] = []

input_files: list[TextIO] = []

RE_PROJECT_START = re.compile(r'^#pragma\s+project\s+"([^"]+)"')
RE_CU_START = re.compile(r'^#pragma\s+echo\s+"Processing (/[^\\]+)\\n"')

def begin_shard(file_number):
    """Start a new output file."""
    outfile = Path(f"file-{file_number:04d}.cs")
    f = outfile.open("w")
    return f

def begin_project(file, name):
    """Start a new project."""
    file.write(f'#pragma echo "Processing project {name}\\n"\n')
    file.write(f'#pragma project "{name}"\n')
    file.write("#pragma block_enter\n")

def end_project(file, name):
    file.write("#pragma block_exit\n")
    file.write(f'#pragma echo "Done processing project {name}\\n\\n"\n')

"""
Algorithm outline
- Go through all specified .cs files keeping the CU offset and project
  for each CU
- Determine CUs per shard
- Go through all files in encountered order
- Allocate each encountered CU and all CUs with the same path to a shard
  until the shard fills up.
- For each shard, for each project, write out the corresponding CUs.
"""
def read_project(input_file):

    global input_files
    global project_id
    global ncus

    input_files.append(input_file)

    # Read properties for each CU
    while True:
        offset = input_file.tell()
        line = input_file.readline()
        if not line:
            break
        line = line.rstrip()

        if matched := RE_PROJECT_START.search(line):
            project_id += 1
            project_name = matched.group(1)
            project_name_to_id[project_name] = project_id
            project_id_to_name[project_id] = project_name
            print(f"Reading project {project_name}")

        elif matched := RE_CU_START.search(line):
            cu_path = matched.group(1)
            ncus += 1
            if cu_path in cu_to_props:
                props = cu_to_props[cu_path]
            else:
                props = CUProps(cu_path)
                cu_to_props[cu_path] = props
            props.add_project(project_id, offset)

            cu_list.append(props)


def allocate(nshards: int):
    global project_id
    global ncus

    nprojects = project_id + 1
    max_shard_cus = ncus / nshards

    # Allocate CUs to shards
    print("Allocating compilation units")
    shard: list[CUProps] = []
    all_shards: list[list[CUProps]] = [shard]
    shard_cus: int = 0
    for cu in cu_list:
        if cu.allocated:
            continue
        shard.append(cu)
        shard_cus += cu.number_of_projects()
        cu.allocated = True
        if shard_cus >= max_shard_cus:
            shard = []
            all_shards.append(shard)
            shard_cus = 0

    # Write out the projects and the CUs in each shard
    shard_number = 0
    for shard in all_shards:
        print(f"Writing out shard {shard_number}")
        output_file = begin_shard(shard_number)
        for project_id in range(0, nprojects):
            project_name = project_id_to_name[project_id]
            print(f"Writing out project {project_id} [{project_name}]")
            begin_project(output_file, project_name)
            for cu in shard:
                if cu.in_project(project_id):
                    cu.write(output_file, project_id)
            end_project(output_file, project_name)
        shard_number += 1

def main():
    """Program entry point"""
    # Ignore unsightly broken pipe exceptions
    signal.signal(signal.SIGPIPE, signal.SIG_DFL)

    parser = argparse.ArgumentParser(
        description='CScout file sharder')
    parser.add_argument('-s', '--shards',
                        help='Number of shards to create',
                        type=int,
                        required=True)
    parser.add_argument('file',
                        help='Files to process',
                        nargs='+',
                        type=str)
    args = parser.parse_args()
    for file_name in args.file:
        input_file = open(file_name)
        read_project(input_file)
    allocate(args.shards)

if __name__ == "__main__":
    main()
